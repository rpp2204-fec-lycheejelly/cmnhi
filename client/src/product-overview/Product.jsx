import React from 'react';
import axios from 'axios';
import ProductInfo from './ProductInfo.jsx';
import ImageGallery from './ImageGallery.jsx';
import StyleSelect from './StyleSelect.jsx';
import AddCart from './AddCart.jsx';
import ProductDesc from './ProductDesc.jsx';


class Product extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      product: {},
      styles: [],
      currentStyle: {},
      sku: null,
      frontIdx: 0,
      backIdx: 7,
      mainIdx: 0,
      matcher: 0,
      carousel: [],
      view: 'default'
    }
  }

  componentDidUpdate(prevProps, prevState) {
    if(JSON.stringify(this.props.productData) !== JSON.stringify(prevProps.productData)) {
      this.setState({
        product: this.props.productData,
        styles: this.props.productData.styles,
        currentStyle: this.props.productData.styles[0]
      })
    }

    if(this.state.mainIdx < 0) {
        let newCarousel = [...this.state.carousel];
        newCarousel.pop();
        newCarousel.unshift(this.state.currentStyle.photos[this.state.frontIdx - 1]);

        this.setState({
          carousel: newCarousel,
          frontIdx: this.state.frontIdx - 1,
          backIdx: this.state.backIdx - 1,
          matcher: this.state.matcher - 1,
          mainIdx: 0
        })

    } else if (this.state.mainIdx > 6) {
        let newCarousel = [...this.state.carousel];
        newCarousel.shift();
        newCarousel.push(this.state.currentStyle.photos[this.state.backIdx]);

        this.setState({
          carousel: newCarousel,
          frontIdx: this.state.frontIdx + 1,
          backIdx: this.state.backIdx + 1,
          matcher: this.state.matcher + 1,
          mainIdx: 6
        })
    }
  }

  changeStyle(style) {
    this.setState({
      currentStyle: style,
      sku: null,
    })
  }

  updateSku(sku) {
    this.setState({
      sku
    })
  }

  addToBag(count) {
    axios.post('/cart', {
      sku_id: this.state.sku,
      count
    })
    .then(result => {
      console.log('Successful POST to /cart')
    })
    .catch(error => {
      console.log('Error POST to /cart', error)
    })
  }

  changeImage(index) {
      this.setState({
        mainIdx: index,
      })
  }

  changeView() {
    this.setState({
      view: 'expanded'
    })
  }

  scrollRight() {
    this.setState({
      mainIdx: this.state.mainIdx + 1,
    })
  }

  scrollLeft() {
    this.setState({
      mainIdx: this.state.mainIdx - 1,
    })
  }

  scrollUp(newCarousel) {
    this.setState({
      carousel: newCarousel,
      frontIdx: this.state.frontIdx + 1,
      backIdx: this.state.backIdx + 1,
      matcher: this.state.matcher + 1
    })
  }

  scrollDown(newCarousel) {
    this.setState({
      carousel: newCarousel,
      frontIdx: this.state.frontIdx - 1,
      backIdx: this.state.backIdx - 1,
      matcher: this.state.matcher -1,
    })
  }

  spliceCarousel(carousel) {
    this.setState({
      carousel: carousel.splice(this.state.frontIdx, 7)
    })
  }

  render() {
    return <>
           <div className='product-overview'>
             <ImageGallery photos={this.state.currentStyle.photos || []}
                           view={this.state.view}
                           frontIdx={this.state.frontIdx}
                           backIdx={this.state.backIdx}
                           mainIdx={this.state.mainIdx}
                           matcher={this.state.matcher}
                           carousel={this.state.carousel}
                           changeView={this.changeView.bind(this)}
                           changeImage={this.changeImage.bind(this)}
                           scrollDown={this.scrollDown.bind(this)}
                           scrollUp={this.scrollUp.bind(this)}
                           scrollLeft={this.scrollLeft.bind(this)}
                           scrollRight={this.scrollRight.bind(this)}
                           spliceCarousel={this.spliceCarousel.bind(this)}
                           />
            {this.state.view === 'default' &&
              <div className='product-information'>
                  <ProductInfo product={this.state.product}
                                reviews={this.state.product.ratings}
                                style={this.state.currentStyle}/>
                  <StyleSelect styles={this.state.styles}
                                currentStyle={this.state.currentStyle}
                                changeStyle={this.changeStyle.bind(this)}/>
                  <AddCart className='add-cart' skus={this.state.currentStyle.skus || {}}
                                                sku={this.state.sku}
                                                updateSku={this.updateSku.bind(this)}
                                                add={this.addToBag.bind(this)}/>
              </div>
            }
          </div>
          <ProductDesc slogan={this.state.product.slogan}
                       description={this.state.product.description}
                       features={this.state.product.features}/>
          </>
  }
}

export default Product

