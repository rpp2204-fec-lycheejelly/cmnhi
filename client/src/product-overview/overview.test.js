import React from 'react';
import renderer from 'react-test-renderer';
import calculateAverage from './lib/calculateAvg.js';
import Product from './Product.jsx';

describe("Simple test suite", function() {
  let sum = (a, b) => {
    return a + b;
  }

  it("Should add 1 + 1", function() {
    expect(sum(1, 1)).toBe(2);
  })
})

describe("calculate Average", function() {
  let reviewData1 = {
    1: 5,
    2: 8,
    3: 9,
    4: 13,
    5: 26
  }
  let reviewData2 = {
    1: 0,
    2: 0,
    3: 0,
    4: 0,
    5: 50,
  }

  it("Should return reviews in specific format", function() {
    expect(calculateAverage(reviewData1).scores).toStrictEqual([1, 1, 1, 0.75, 0]);
    expect(calculateAverage(reviewData2).scores).toStrictEqual([1, 1, 1, 1, 1]);
  })

  it("Should calculate total reviews", function() {
    expect(calculateAverage(reviewData1).totalResponses).toBe(61);
    expect(calculateAverage(reviewData2).totalResponses).toBe(50);
  })
})

describe("Component Testing", function() {
  it("Should correctly render productInfo", function() {
    const tree = renderer.create(
      <Product />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  })
})