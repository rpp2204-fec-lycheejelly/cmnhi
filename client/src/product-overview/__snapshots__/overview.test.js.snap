// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Testing Should correctly render productInfo 1`] = `
<div
  className="product-overview"
>
  <div
    className="image-gallery"
  >
    <div
      className="image-default-view"
    />
  </div>
  <div
    className="product-information"
  >
    <div>
      <div
        className="stars"
      >
        <div
          className="star-container"
        >
          <div
            className="star-fill"
            style={
              Object {
                "width": "0px",
              }
            }
          >
            <img
              alt="star alt"
              className="star-outline"
              src=""
            />
          </div>
        </div>
        <div
          className="star-container"
        >
          <div
            className="star-fill"
            style={
              Object {
                "width": "0px",
              }
            }
          >
            <img
              alt="star alt"
              className="star-outline"
              src=""
            />
          </div>
        </div>
        <div
          className="star-container"
        >
          <div
            className="star-fill"
            style={
              Object {
                "width": "0px",
              }
            }
          >
            <img
              alt="star alt"
              className="star-outline"
              src=""
            />
          </div>
        </div>
        <div
          className="star-container"
        >
          <div
            className="star-fill"
            style={
              Object {
                "width": "0px",
              }
            }
          >
            <img
              alt="star alt"
              className="star-outline"
              src=""
            />
          </div>
        </div>
        <div
          className="star-container"
        >
          <div
            className="star-fill"
            style={
              Object {
                "width": "0px",
              }
            }
          >
            <img
              alt="star alt"
              className="star-outline"
              src=""
            />
          </div>
        </div>
        <a>
          Read All 
          0
           Reviews
        </a>
      </div>
      <p />
      <h4 />
      <span
        className=""
      >
        $
         
      </span>
      <p
        className="product-description"
      />
    </div>
    <div
      className="style-select"
    >
      <p>
         
        <b>
          Style &gt;
        </b>
         
      </p>
      <div
        className="style-container"
      />
    </div>
    <div>
      <select>
        <option>
          Select Size
        </option>
        <option>
          XS
        </option>
        <option>
          S
        </option>
        <option>
          M
        </option>
        <option>
          L
        </option>
        <option>
          XL
        </option>
      </select>
      <select>
        <option>
          1
        </option>
        <option>
          2
        </option>
        <option>
          3
        </option>
        <option>
          4
        </option>
        <option>
          5
        </option>
      </select>
      <button>
        Add To Cart
      </button>
      <button>
        â˜†
      </button>
    </div>
  </div>
</div>
`;
